design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,156 @@
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+parameters:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+The resource claim below specifies that it is of `kind: MySQLInstance`, and references the above resource class in its `classRef`.
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  classRef:
+    name: standard-mysql
+    namespace: crossplane-system
+  engineVersion: "9.6"
+```
+
+If using a default resource class, the class and claim would be adjusted as follows.
+
+Resource Class:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+  labels:
+    mysqlinstance.storage.crossplane.io/default: "true"
+parameters:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+Resource Claim:
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  engineVersion: "9.6"
+```
+
+This model is powerful because it allows an application developer to create a resource claim without having to know the implementation details or even the underlying provider. However, the fact that every resource class if of the same `kind` presents a key issue: The required parameters for a resource class may vary widely, and they are currently only provided as an arbitrary map that is eventually read by the controller for the the specified `provisioner`. Therefore, an administrator who is creating resource classes does not know what fields are required and will not be notified of missing or extraneous fields until the provisioning of a resource that references the class.
+
+This problem can be solved by the implementation of "strongly typed" resource classes. Instead of every resource class being of `kind: ResourceClass`, they would map one-to-one with the concrete managed resource kind for which they provided specifications (i.e. `kind: RDSInstanceClass`).
+
+Unfortunately, this makes it much more difficult to identify default resource classes because the default class controller currently watches for objects of `kind: ResourceClass`. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller cannot determine all class kinds to watch.
```

Comment:
I agree that it would be good to specify there is one default controller per claim type. Also, I need to actually fix this entire paragraph because default class controllers *watch* for objects of their resource claim type, and *list* objects of `kind: ResourceClass`.